content/en/blog/2024/otel-generative-ai/index.md

```diff
@@ -0,0 +1,204 @@
+example, temperature, top_p) and response details like token count or errors.
+They provide visibility into each request, aiding in identifying bottlenecks and
+analyzing the impact of settings on model output.
+
+### Metrics: Monitoring Usage and Performance
+
+Metrics aggregate high-level indicators like request volume, latency, and token
+counts, essential for managing costs and performance. This data is particularly
+critical for API-dependent AI applications with rate limits and cost
+considerations.
+
+### Events: Capturing Detailed Interactions
+
+Events log detailed moments during model execution, such as user prompts and
+model responses, providing a granular view of model interactions. These insights
+are invaluable for debugging and optimizing AI applications where unexpected
+behaviors may arise.
+
+{{% alert title="Note" color="info" %}} Note that we decided to use the newer
+[Log events API](https://opentelemetry.io/docs/specs/otel/logs/event-api/)
+specification in the Semantic Conventions for Generative AI. The events API
+allows for us to define specific
+[semantic conventions](https://opentelemetry.io/docs/specs/semconv/general/events/)
+for the user prompts and model responses that we capture. {{% /alert %}}
+
+### Extending Observability with Vendor-Specific Attributes
+
+The Semantic Conventions also define vendor-specific attributes for platforms
+like OpenAI and Azure Inference API, ensuring telemetry captures both general
+and provider-specific details. This added flexibility supports multi-platform
+monitoring and in-depth insights.
+
+## Building the Python Instrumentation Library for OpenAI
+
+This Python-based library for OpenTelemetry captures key telemetry signals for
+OpenAI models, providing developers with an out-of-the-box observability
+solution tailored to AI workloads. The library,
+[hosted within the OpenTelemetry Python Contrib repository](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/opentelemetry-instrumentation-openai-v2%3D%3D2.0b0/instrumentation-genai/opentelemetry-instrumentation-openai-v2),
+automatically collects telemetry from OpenAI model interactions, including
+request and response metadata and token usage.
+
+As generative AI applications grow, additional instrumentation libraries for
+other languages will follow, extending OpenTelemetry support across more tools
+and environments. The current library’s focus on OpenAI highlights its
+popularity and demand within AI development, making it a valuable initial
+implementation.
+
+### Example Usage
+
+Here’s an example of using the OpenTelemetry Python library to monitor a
+generative AI application with the OpenAI client. Make sure you first install
+the library:
+
+```bash
+pip install opentelemetry-instrumentation-openai-v2
+```
+
+Set the following environment variables, updating the endpoint and protocol as
+appropriate:
+
+```bash
+OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
```

Comment:
TL;DR; IMHO whatever we paste in a bash codeblock, people will expect to work by pasting on their own. This doesn't work, yet even if we fix the pip deps due to not exporting the variables.

maybe worth removing these two `OTEL_EXPORTER` variables and leave it to defaults based on the dependency like `opentelemetry-exporter-otlp-proto-grpc` or `opentelemetry-exporter-otlp-proto-http`. wdyt?

Then, we have only the service name, api key and capture content ones. Also, these need to be exported or set inline. If you don't the python command won't work later. Many people put these into a .env file and use dotenv to load it.

e.g. 

```bash
$ pip install "python-dotenv[cli]"
$ dotenv run -- python main.py
```

However, we can also just inline them like

```bash
$ OTEL_SERVICE_NAME=python-opentelemetry-openai OPENAI_API_KEY=sk-not_tellin OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT=true python main.py
```
```